docs/DesignDocs/Enabledebugmalloc.md

```diff
@@ -0,0 +1,229 @@
+several pull requests, then this doc will emphasize the design of "local
+detection".
+
+# User Experience
+
+## Global detection
+
+### Default option
+
+The default option is "USE_DEBUG_MALLOC". Now we already set "USE_DEBUG_MALLOC"
+to "ON" in debug mode by default. Users can modify this default option
+according to the need, such as change it to "ON" in release mode or change it
+to "OFF" in debug mode.
+
+### Customed option
+
+Debug malloc will be shipped to the users as a separate library
+oelibdebugmalloc.a, so users can turn on debug malloc per enclave while the
+default option is "OFF".
+
+To use debug malloc, users will have to link oelibdebugmalloc.a into their
+enclaves explictly. This can be done by
+```
+      cmake:
+      target_link_libraries(user-enclave oelibc oedebugmalloc oeenclave)
+
+      make:
+      LDFLAGS=$(shell pkg-config oeenclave-$(COMPILER) --libs)
+      LDFLAGS_CUSTOM=$(subst -loelibc, -loedebugmalloc -loelibc, $(LDFLAGS))
+      $(CC) -o user-enclave user-enc.o $(LDFLAGS_CUSTOM)
+
+```
+
+For the global detection, if there is any memory leak during the enclave's
+life circle, an error code "OE_MEMORY_LEAK" will be raised during the enclave
+termination. This means it not only can check the users' enclave application,
+but also can cover the SDK's allocation of the enclave side.
+
+## Local detection
+
+Currently, debug malloc prints leaks only during enclave termination.
+However, users would like to isolate and track specific memory allocations.
+To achieve this, the following APIs are provided in a new published header `debugmalloc.h`.
+
+- `bool oe_debug_malloc_tracking_start(void)`
+  Start tracking memory allocations.
+
+- `bool oe_debug_malloc_tracking_stop(void)`
+  Stop tracking memory allocations.
+
+- `int oe_debug_malloc_tracking_report(void)`
+  report the number of tracked objects that alive, ie not yet freed.
+
+Please note that the timing and sequences of calling of these functions is
+important:
+
+- The starting and stopping tracking must cover the coupled malloc/free
+operations, which means start tracking before malloc and also stop tracking
+after free, or the report will definitely trigger false alarms.
+
+- oe_debug_malloc_tracking_stop() can only been called after calling
+oe_debug_malloc_tracking_start(), and oe_debug_malloc_tracking_start() can be
+called at the beginning or after calling oe_debug_malloc_tracking_stop(). While
+the calling sequence following this rule, the return values are false. TRUE, or
+non-zero return value means the sequence is wrong and the report does not
+reflect the real situation.
+
+- oe_debug_malloc_tracking_start() and oe_debug_malloc_tracking_stop() can been
+called many times:
+
+```
+  |----code1--|-----------|----code2--|--|----code3--|
+  start1->stop1->report1->start2->stop2->start3->stop3->report3
+```
+
+In this case report1 only covers code1, and report3 covers code1, code2 and
```

Comment:
Updated, thanks!

Reply:
This cumulative tracking actually surprises me, and is not what I would expect looking at the state diagram that you added. Why are we tracking across `stop` boundaries? I would expect each start-stop pair to reset the state of local tracking. 

+@anakrish for his opinion on this.

Reply:
Considering tracking state and not tracking state as an entirety respectively, `stop` is not a boundary. Of course this is my subjective opinion, this design is based on such case:  since calling start() and stop() is easier than report(), the user can define many code segments as tracking state, then one time calling report() is enough to cover all the involved segments.